qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        return Signal(envelope=merged_env, carrier_freq=ave_freq, name=str(self))
 
-    Returns:
-        signal: The type will depend on the given base class.
 
-    Raises:
-        Exception: if signals cannot be added.
+class DiscreteSignalSum(DiscreteSignal, SignalSum):
+    """Represents a sum of piecewise constant signals, all with the same
+    time parameters: dt, number of samples, and start time.
     """
 
-    # ensure both arguments are signals
-    if isinstance(sig1, (int, float, complex)):
-        sig1 = Constant(sig1)
-    if isinstance(sig2, (int, float, complex)):
-        sig2 = Constant(sig2)
+    def __init__(
+        self,
+        dt: float,
+        samples: Union[Array, List],
+        start_time: float = 0.0,
+        duration: int = None,
+        carrier_freqs: float = None,
+        phases: float = None,
+        name: str = None,
+    ):
+        """Samples array has 0th axis corresponding to a signal, 1st axis corresponding to samples
+        for each signal."""
+        self._name = name
+        self._dt = dt
+        self._samples = Array(samples)
+        self._start_time = start_time
 
-    return Signal(
-        lambda t: sig1.value(t) + sig2.value(t),
-        0,
-        0,
-    )
+        if carrier_freqs is None:
+            carrier_freqs = np.zeros(len(samples), dtype=float)
+
+        if phases is None:
+            phases = np.zeros(len(samples), dtype=float)
+
+        # construct individual components so they can be accessed as in SignalSum
+        components = []
+        for sample_row, freq, phase in zip(self.samples.transpose(), carrier_freqs, phases):
+            components.append(
+                DiscreteSignal(
+                    dt=self.dt,
+                    samples=sample_row,
+                    start_time=self.start_time,
+                    duration=self.duration,
+                    carrier_freq=freq,
+                    phase=phase,
+                )
+            )
 
+        self.components = components
 
-class VectorSignal:
-    """The vector version of the Signal class - the envelope is an array-valued
-    function, and carrier_freqs is an array of carrier frequencies.
+        # initialize internally stored carrier/phase information
+        self._carrier_freq = None
+        self._phase = None
+        self._carrier_arg = None
+        self._phase_arg = None
 
-    In addition, a drift_array is set to correspond to the value of the
-    VectorSignal when all "time-dependent terms" are off. E.g. if it is
-    composed of a list of Signal objects, this corresponds to the output when
-    all non-Constant signal objects are zero.
-    """
+        self.carrier_freq = carrier_freqs
+        self.phase = phases
 
-    def __init__(
-        self,
-        envelope: Callable,
-        carrier_freqs: Array,
-        phases: Array,
-        drift_array: Optional[Array] = None,
+    @classmethod
+    def from_SignalSum(
```

Comment:
Does this need to do any validation to check the assumptions of DiscreteSignal sum or does it enforce them by sampling the given `SignalSum`? Perhaps this can be clarified in the docstring?